Add theory, intro and use case pages of LRE user guide (#2522)
unitaryfund/mitiq: docs/source/guide/lre-5-theory.md

---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.11.4
kernelspec:
  display_name: Python 3
  language: python
  name: python3
---

> ```{admonition} Warning:
> The user guide for LRE in Mitiq is currently under construction.
> ```

Reviewer: ````suggestion
```{warning}
The user guide for LRE in Mitiq is currently under construction.
```
````

# What is the theory behind LRE?

Layerwise Richardson Extrapolation (LRE), an error mitigation technique, introduced in
{cite}`Russo_2024_LRE` works by creating multiple noise-scaled variations of the input
circuit such that the noiseless expectation value is extrapolated from the execution of each

> noisy circuit.

Reviewer: I think this description is too easily confusable with ZNE. Can we say something like the following?

```md
LRE is an EM strategy which extends ideas found in ZNE by allowing users to create many noise-scaled circuits by folding individual layers instead of the entire circuit.
```

It should use some more tweaking.

Similar to [ZNE](zne.md), this process works in two steps:

> - **Step 1:** Intentionally create multiple noise-scaled but logically equivalent circuits through unitary folding.

Author: I am going to skip using the word 'layerwise' as we have a similar submodule in ZNE. There a user can choose which layers they want to fold.

> - **Step 1:** Intentionally create multiple noise-scaled but logically equivalent circuits through unitary folding.

Reviewer: Makes sense! I still think it might be nice to find a way to throw something in here to help someone understand &mdash; even if it's from a very high-level perspective &mdash; how these circuits are different from noise-scaled circuits in ZNE.

> - **Step 1: Intentionally create multiple noise-scaled but logically equivalent circuits** through unitary folding.

Reviewer 2: I wonder if the `**` delimiters should be around `Step 1:` and `Step 2:` (instead of around the larger chunk of text that they are now surrounding.

LRE leverages the flexible configurational space of layerwise unitary folding,
allowing for a more nuanced mitigation of errors by treating the noise level of each layer of

> the quantum circuit as an independent variable

Reviewer 2: [nit]: Period at end of sentence.

## Step 1: Intentionally create multiple noise-scaled but logically equivalent circuits

The goal is to create noise-scaled circuits of different depths where the layers in each circuit are scaled in
a specific pattern as a result of unitary folding. This pattern is often described by the vector of scale factor vectors
generated by the fold multiplier and the chosen degree for multivariate Richardson extrapolation polynomial. For more information
on unitary folding, go to [What is the theory behind ZNE?](zne-5-theory.md).

Suppose we want to estimate the noiseless expectation value of some observable in an $n$-qubit circuit with $l$ layers.

> Each layer can have a different scale factor and we can create $M$ such variations of the scaled circuit. Let $\{λ_1, λ_2, λ_3, ..., λ_M\}$ be the scale factors vectors used to create multiple variations of the noise-scaled circuits $\{C_{λ_1}, C_{λ_2}, C_{λ_3}, ……, C_{λ_M}\}$ such that each vector $λ_i$ defines the scale factors for the different layers in the input circuit $\{{λ^1}_i, {λ^2}_i, {λ^3}_i, ….., {λ^l}_i\}^T$.

Reviewer 2: Looks like the continuation dots here are all different. I think for all of these we want `\ldots`.

> If $d$ is the chosen degree of our multivariate polynomial, $M_j(λ_i, d)$ corresponds to the terms in the polynomial. In general, the monomial terms for a variable $l$ up to degree $d$  can be determined through the [Stars and Bars method](https://en.wikipedia.org/wiki/Stars_and_bars_%28combinatorics%29).

Reviewer 2: [nit]: Extra space between $d$ and "can".

> If $d$ is the chosen degree of our multivariate polynomial, $M_j(λ_i, d)$ corresponds to the terms in the polynomial. In general, the monomial terms for a variable $l$ up to degree $d$  can be determined through the [Stars and Bars method](https://en.wikipedia.org/wiki/Stars_and_bars_%28combinatorics%29).

Reviewer 2: I don't think "Stars and Bars" is a proper noun. That is, I think it should be "stars and bars method".

$$
\text{total number of terms in the monomial basis with max degree } d = \binom{d + l}{d}
$$

$$
\text{number of terms in the monomial basis with total degree } d = \binom{d + l - 1}{d}
$$

These monomial terms define the rows of the square sample matrix where $a_{i,j}=M_j(λ_i, d)$.

$$
\mathbf{A}(\Lambda, d) = 
\begin{bmatrix}
    a_{1,1} & a_{1,2} & \cdots & a_{1,M} \\
    a_{2,1} & a_{2,2} & \cdots & a_{2,M} \\
    \vdots & \vdots & \ddots & \vdots \\
    a_{N,1} & a_{N,2} & \cdots & a_{N,M}
\end{bmatrix}
$$

> Each monomial term in the sample matrix is evaluated using the values in the scale factor vectors. We aim to define the zero-noise limit as a linear combination of the noisy expectation values. Finding the coefficients in the linear combination becomes a problem solvable through a system of linear equations $Ac = z$ where c is the coefficients vector, $z$ is the vector of expectation values and $\mathbf{A}$ is the sample matrix evaluated using the values in the scale factor vectors.

Reviewer 2: Missing `$` delimiters around `c`.

## Step 2: Extrapolate to the noiseless limit

> Each noise scaled circuit $C_{λ_i}$ has an expectation value associated with it $\langle O(λ_i) \rangle$ such that we can define a vector of the noisy expectation values $z = (\langle O(λ_1) \rangle, \langle O(λ_2) \rangle, \langle O(λ_3) \rangle, ..., \langle O(λ_M)\rangle)^T$. These have a coefficient of linear combination associated with them such that 

Reviewer 2: Should use `\ldots` here.

$$

> 

Reviewer 2: [nit]: Extra spacing between `$$` delimiters and the equation.

O_{LRE} = \sum_{i=1}^{M} \eta_i \langle O(\boldsymbol{\lambda}_i) \rangle

$$

The system of linear equations is used to find the numerous $\eta_i$. As we only need to find the noiseless expectation value, we do not need to calculate the full vector of linear combination coefficients if we use the [Lagrange interpolation formula](https://files.eric.ed.gov/fulltext/EJ1231189.pdf). 

$$

> O_{\rm LRE} = \sum_{i=1}^M \langle O (\boldsymbol{\lambda}_i)\rangle  \frac{\det \left(\mathbf{M}_i (\boldsymbol{0}) \right)}{\det \left(\mathbf{A}\right)}

Reviewer 2: [nit]: Period at end of equation for sentence termination.

> O_{\rm LRE} = \sum_{i=1}^M \langle O (\boldsymbol{\lambda}_i)\rangle  \frac{\det \left(\mathbf{M}_i (\boldsymbol{0}) \right)}{\det \left(\mathbf{A}\right)}.

Author: This is supposed to be one of the terms in the monomial basis evaluated at $\lambda=0$

> O_{\rm LRE} = \sum_{i=1}^M \langle O (\boldsymbol{\lambda}_i)\rangle  \frac{\det \left(\mathbf{M}_i (\boldsymbol{0}) \right)}{\det \left(\mathbf{A}\right)}.

Reviewer: Why not just use $M_i(\boldsymbol{0}, d)$? That's the notation defined above. Also since we defined $M$ with a second argument, I don't think we should drop it.

$$

> To get the matrix $\mathbf{M}_i(\mathbf{0})$, replace the $i$-th row of the sample matrix $\mathbf{A}$ by $\mathbf{e}_1=(1, 0, \ldots, 0)^t$.

Reviewer 2: Sometimes `^t` or `^T` is used to denote the transpose. Should use one universally (say, `^T`).

circuit such that the noiseless expectation value is extrapolated from the execution of each
noisy circuit.

Similar to [ZNE](zne.md), this process works in two steps:

- **Step 1:** Intentionally create multiple noise-scaled but logically equivalent circuits by scaling each layer or chunk of the input circuit through unitary folding.

- **Step 2:** Extrapolate to the noiseless limit using multivariate richardson extrapolation.

LRE leverages the flexible configuration space of layerwise unitary folding,
allowing for a more nuanced mitigation of errors by treating the noise level of each layer of
the quantum circuit as an independent variable.

## Step 1: Intentionally create multiple noise-scaled but logically equivalent circuits

The goal is to create noise-scaled circuits of different depths where the layers in each circuit are scaled in
a specific pattern as a result of unitary folding. This pattern is often described by the vector of scale factor vectors
generated by the fold multiplier and the chosen degree for multivariate Richardson extrapolation polynomial. For more information

> on unitary folding, go to [What is the theory behind ZNE?](zne-5-theory.md).

Author: > Add a bit more of a lead in. An example where we have a simple circuit, and we show what each λ i is would be super helpful.

I have this description in part 4 of the lre user guide (not pushed yet because that's what I am currently working on).

> on unitary folding, go to [What is the theory behind ZNE?](zne-5-theory.md).

Reviewer: Ah great! In that case it probably makes sense to add a link to that section/example from this part.

Suppose we're interested in the value of some observable in an $n$-qubit circuit with $l$ layers.

> Each layer can have a different scale factor and we can create $M$ such variations of the scaled circuit. Let $\{λ_1, λ_2, λ_3, \ldots, λ_M\}$ be the scale factors vectors used to create multiple variations of the noise-scaled circuits $\{C_{λ_1}, C_{λ_2}, C_{λ_3}, \ldots, C_{λ_M}\}$ such that each vector $λ_i$ defines the scale factors for the different layers in the input circuit $\{{λ^1}_i, {λ^2}_i, {λ^3}_i, \ldots, {λ^l}_i\}^T$.

Reviewer: 1. Can we make it a bit more clear here what is a set and what is a vector? E.g. set notation is used in conjunction with a transpose.
2. I think $M$ is fixed in the beginning, right? As it reads now, it sounds like a bound on the number of scaled circuits we can create with this method. IIRC, $M$ is set **by** the user. I might be misunderstanding, though.
3. Just checking my understanding: $C_{\lambda_i}$ represents a single circuit, right? And in order for this to work we should have `len(C) == len(λ_i)`.

Author: >  M is set by the user.

Based on user's choice of degree of the extrapolating polynomial.

Author: > Just checking my understanding: C λ i represents a single circuit, right? And in order for this to work we should have len(C) == len(λ_i)

yes

> If $d$ is the chosen degree of our multivariate polynomial, $M_j(λ_i, d)$ corresponds to the terms in the polynomial arranged in increasing order. In general, the monomial terms for a variable $l$ up to degree $d$ can be determined through the [stars and bars method](https://en.wikipedia.org/wiki/Stars_and_bars_%28combinatorics%29).

Reviewer: ```suggestion
If $d$ is the chosen degree of our multivariate polynomial, we define $M_j(λ_i, d)$ to be the terms in the polynomial arranged in increasing order. In general, the number of monomial terms with $l$ variables up to degree $d$ can be determined through the [stars and bars method](https://en.wikipedia.org/wiki/Stars_and_bars_%28combinatorics%29).
```

$$
\text{total number of terms in the monomial basis with max degree } d = \binom{d + l}{d}
$$

$$
\text{number of terms in the monomial basis with total degree } d = \binom{d + l - 1}{d}

> $$

Reviewer: Can you motivate having this here? It doesn't seem like it's used at any further point.

These monomial terms define the rows of the square sample matrix as shown below:

$$

> \mathbf{A}(\Lambda, d) = 

Reviewer: I don't think $\Lambda$ has been defined. I assume it's $\Lambda = [\lambda_1, \ldots, \lambda_M]^T$, but we shouldn't leave anything up to guessing.

\begin{bmatrix}
    M_1(λ_1, d) & M_2(λ_1, d) & \cdots & M_N(λ_1, d) \\
    M_1(λ_2, d) & M_2(λ_2, d) & \cdots & M_N(λ_2, d) \\
    \vdots & \vdots & \ddots & \vdots \\
    M_1(λ_N, d) & M_2(λ_N, d) & \cdots & M_N(λ_N, d)
\end{bmatrix}
$$

Each monomial term in the sample matrix $\mathbf{A}$ is evaluated using the values in the scale factor vectors. In Step 2, we aim to define $O_{\mathrm{LRE}}$ as a linear combination of the noisy expectation values.

> Finding the coefficients in the linear combination becomes a problem solvable through a system of linear equations $\mathbf{A} c = z$ where $c$ is the coefficients vector $(\eta_1, \eta_2, \ldots, \eta_N)^T$, $z$ is the vector of the noisy expectation values and $\mathbf{A}$ is the sample matrix evaluated using the values in the scale factor vectors.

Reviewer: You mention linear combination before it has been defined or motivated. It feels backward to me.

## Step 2: Extrapolate to the noiseless limit

> Each noise scaled circuit $C_{λ_i}$ has an expectation value $\langle O(λ_i) \rangle$ associated with it such that we can define a vector of the noisy expectation values $z = (\langle O(λ_1) \rangle, \langle O(λ_2) \rangle, \ldots, \langle O(λ_M)\rangle)^T$. These have a coefficient of linear combination associated with them as shown below: 

Reviewer: How do we motivate combining these results in a linear combination?
```suggestion
Each noise scaled circuit $C_{λ_i}$ has an expectation value $\langle O(λ_i) \rangle$ associated with it such that we can define a vector of the noisy expectation values $z = (\langle O(λ_1) \rangle, \langle O(λ_2) \rangle, \ldots, \langle O(λ_M)\rangle)^T$.
These values can then be combined via a linear combination to estimate the ideal value $variable$.
```

$$
O_{\mathrm{LRE}} = \sum_{i=1}^{M} \eta_i \langle O(λ_i) \rangle.
$$

The system of linear equations is used to find the numerous $\eta_i$ in vector $c$. As we only need to find the noiseless expectation value, we can skip calculating the full vector of linear combination coefficients if we use the [Lagrange interpolation formula](https://files.eric.ed.gov/fulltext/EJ1231189.pdf) evaluated at $λ = 0$.

$$
O_{\rm LRE} = \sum_{i=1}^M \langle O (\boldsymbol{\lambda}_i)\rangle  \frac{\det \left(\mathbf{M}_i (\boldsymbol{0}) \right)}{\det \left(\mathbf{A}\right)}.
$$

> To get the matrix $\mathbf{M}_i(\mathbf{0})$, replace the $i$-th row of the sample matrix $\mathbf{A}$ by $\mathbf{e}_1=(1, 0, \ldots, 0)^T$ where except $M_1(0, d) = 1$ all the other monomial terms are zero.

Reviewer: 1. Where does this come from?
2. Conclusion is needed. Recap what we went through, and maybe add a link to the API-doc LRE section?

Author: > Where does this come from?

Discussed before the equation.
